packages/preview2-shim/lib/io/worker-thread.js

```diff
@@ -0,0 +1,312 @@
+import { runAsWorker } from '../synckit/index.js';
+import * as calls from './calls.js';
+import * as streamTypes from './stream-types.js';
+import * as fs from 'node:fs';
+
+let streamCnt = 0, pollCnt = 0;
+
+/** @type {Map<number, Promise<void>>} */
+const unfinishedPolls = new Map();
+
+/** @type {Map<number, { flushPromise: Promise<void> | null, stream: NodeJS.ReadableStream | NodeJS.WritableStream, blocksMainThread: bool }>} */
+const unfinishedStreams = new Map();
+
+/**
+ * 
+ * @param {number} streamId 
+ * @param {NodeJS.ReadableStream | NodeJS.WritableStream} stream 
+ */
+function streamError (streamId, stream, err) {
+  if (stream.end)
+    stream.end();
+  // we delete the stream from unfinishedStreams as it is now "finished" (closed)
+  unfinishedStreams.delete(streamId);
+  return {
+    tag: 'stream-error',
+    val: { tag: 'last-operation-failed', val: err }
+  };
+}
+
+/**
+ * 
+ * @param {number} streamId 
+ * @returns {{ stream: NodeJS.ReadableStream | NodeJS.WritableStream, flushPromise: Promise<void> | null, blocksMainThread: bool }}
+ */
+function getStreamOrThrow (streamId) {
+  const stream = unfinishedStreams.get(streamId);
+  // not in unfinished streams <=> closed
+  if (!stream)
+    throw { tag: 'stream-error', val: { tag: 'closed' } };
+  if (stream.stream.errored)
+    throw streamError(streamId, stream, stream.stream.errored);
+  if (stream.stream.closed) {
+    unfinishedStreams.delete(streamId);
+    throw { tag: 'closed' };
+  }
+  return stream;
+}
+
+/**
+ * @param {number} call
+ * @param {number | null} id
+ * @param {any} payload
```

Comment:
what's the type of `payload`?

Reply:
Because this function represents every possible call, payload will be whatever is needed for the specific call. It is therefore very specifically `any`.